lib/samtools/htslib-1.3.1/tabix.1
.TH tabix 1 "22 April 2016" "htslib-1.3.1" "Bioinformatics tools"
.SH NAME
.PP
bgzip \- Block compression/decompression utility
.PP
tabix \- Generic indexer for TAB-delimited genome position files
.\"
.\" Copyright (C) 2009-2011 Broad Institute.
.\"
.\" Author: Heng Li <lh3@sanger.ac.uk>
.\"
.\" Permission is hereby granted, free of charge, to any person obtaining a
.\" copy of this software and associated documentation files (the "Software"),
.\" to deal in the Software without restriction, including without limitation
.\" the rights to use, copy, modify, merge, publish, distribute, sublicense,
.\" and/or sell copies of the Software, and to permit persons to whom the
.\" Software is furnished to do so, subject to the following conditions:
.\"
.\" The above copyright notice and this permission notice shall be included in
.\" all copies or substantial portions of the Software.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
.\" IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
.\" FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
.\" THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
.\" LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
.\" FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
.\" DEALINGS IN THE SOFTWARE.
.\"
.SH SYNOPSIS
.PP
.B bgzip
.RB [ -cdhB ]
.RB [ -b
.IR virtualOffset ]
.RB [ -s
.IR size ]
.RI [ file ]
.PP
.B tabix
.RB [ -0lf ]
.RB [ -p
gff|bed|sam|vcf]
.RB [ -s
.IR seqCol ]
.RB [ -b
.IR begCol ]
.RB [ -e
.IR endCol ]
.RB [ -S
.IR lineSkip ]
.RB [ -c
.IR metaChar ]
.I in.tab.bgz
.RI [ "region1 " [ "region2 " [ ... "]]]"

.SH DESCRIPTION
.PP
Tabix indexes a TAB-delimited genome position file
.I in.tab.bgz
and creates an index file (
.I in.tab.bgz.tbi
or 
.I in.tab.bgz.csi
) when
.I region
is absent from the command-line. The input data file must be position
sorted and compressed by
.B bgzip
which has a
.BR gzip (1)
like interface. After indexing, tabix is able to quickly retrieve data
lines overlapping
.I regions
specified in the format "chr:beginPos-endPos". Fast data retrieval also
works over network if URI is given as a file name and in this case the
index file will be downloaded if it is not present locally.

.SH INDEXING OPTIONS
.TP 10
.B -0, --zero-based
Specify that the position in the data file is 0-based (e.g. UCSC files)
rather than 1-based.
.TP
.BI "-b, --begin " INT
Column of start chromosomal position. [4]
.TP
.BI "-c, --comment " CHAR
Skip lines started with character CHAR. [#]
.TP
.BI "-C, --csi"
Produce CSI format index instead of classical tabix or BAI style indices.
.TP
.BI "-e, --end " INT
Column of end chromosomal position. The end column can be the same as the
start column. [5]
.TP
.B "-f, --force "
Force to overwrite the index file if it is present.
.TP
.BI "-m, --min-shift" INT
set minimal interval size for CSI indices to 2^INT [14]
.TP
.BI "-p, --preset " STR
Input format for indexing. Valid values are: gff, bed, sam, vcf.
This option should not be applied together with any of
.BR -s ", " -b ", " -e ", " -c " and " -0 ;
it is not used for data retrieval because this setting is stored in
the index file. [gff]
.TP
.BI "-s, --sequence " INT
Column of sequence name. Option
.BR -s ", " -b ", " -e ", " -S ", " -c " and " -0
are all stored in the index file and thus not used in data retrieval. [1]
.TP
.BI "-S, --skip-lines " INT
Skip first INT lines in the data file. [0]

.SH QUERYING AND OTHER OPTIONS
.TP
.B "-h, --print-header "
Print also the header/meta lines.
.TP
.B "-H, --only-header "
Print only the header/meta lines.
.TP
.B "-l, --list-chroms "
List the sequence names stored in the index file.
.TP
.B "-r, --reheader " FILE
Replace the header with the content of FILE
.TP
.B "-R, --regions " FILE
Restrict to regions listed in the FILE. The FILE can be BED file (requires .bed, .bed.gz, .bed.bgz 
file name extension) or a TAB-delimited file with CHROM, POS, and,  optionally,
POS_TO columns, where positions are 1-based and inclusive.  When this option is in use, the input
file may not be sorted. 
regions.
.TP
.B "-T, --targets" FILE
Similar to 
.B -R
but the entire input will be read sequentially and regions not listed in FILE will be skipped.
.PP
.SH EXAMPLE
(grep ^"#" in.gff; grep -v ^"#" in.gff | sort -k1,1 -k4,4n) | bgzip > sorted.gff.gz;

tabix -p gff sorted.gff.gz;

tabix sorted.gff.gz chr1:10,000,000-20,000,000;

.SH NOTES
It is straightforward to achieve overlap queries using the standard
B-tree index (with or without binning) implemented in all SQL databases,
or the R-tree index in PostgreSQL and Oracle. But there are still many
reasons to use tabix. Firstly, tabix directly works with a lot of widely
used TAB-delimited formats such as GFF/GTF and BED. We do not need to
design database schema or specialized binary formats. Data do not need
to be duplicated in different formats, either. Secondly, tabix works on
compressed data files while most SQL databases do not. The GenCode
annotation GTF can be compressed down to 4%.  Thirdly, tabix is
fast. The same indexing algorithm is known to work efficiently for an
alignment with a few billion short reads. SQL databases probably cannot
easily handle data at this scale. Last but not the least, tabix supports
remote data retrieval. One can put the data file and the index at an FTP
or HTTP server, and other users or even web services will be able to get
a slice without downloading the entire file.

.SH AUTHOR
.PP
Tabix was written by Heng Li. The BGZF library was originally
implemented by Bob Handsaker and modified by Heng Li for remote file
access and in-memory caching.

.SH SEE ALSO
.PP
.BR samtools (1)


lib/bcftools/tabix.c
/*  tabix.c -- tabix subcommand.

    Copyright (C) 2012 Broad Institute.
    Copyright (C) 2013 Genome Research Ltd.

    Author: Heng Li <lh3@sanger.ac.uk>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.  */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <htslib/bgzf.h>
#include <htslib/tbx.h>

int main_tabix(int argc, char *argv[])
{
    int c, min_shift = -1, is_force = 0, is_all = 0;
    tbx_conf_t conf = tbx_conf_gff, *conf_ptr = NULL;
    while ((c = getopt(argc, argv, "0fap:s:b:e:S:c:m:")) >= 0)
        if (c == '0') conf.preset |= TBX_UCSC;
        else if (c == 'f') is_force = 1;
        else if (c == 'a') is_all = 1;
        else if (c == 'm') min_shift = atoi(optarg);
        else if (c == 's') conf.sc = atoi(optarg);
        else if (c == 'b') conf.bc = atoi(optarg);
        else if (c == 'e') conf.ec = atoi(optarg);
        else if (c == 'c') conf.meta_char = *optarg;
        else if (c == 'S') conf.line_skip = atoi(optarg);
        else if (c == 'p') {
            if (strcmp(optarg, "gff") == 0) conf_ptr = &tbx_conf_gff;
            else if (strcmp(optarg, "bed") == 0) conf_ptr = &tbx_conf_bed;
            else if (strcmp(optarg, "sam") == 0) conf_ptr = &tbx_conf_sam;
            else if (strcmp(optarg, "vcf") == 0) conf_ptr = &tbx_conf_vcf;
            else {
                fprintf(stderr, "The type '%s' not recognised\n", optarg);
                return 1;
            }

        }
    if (optind == argc) {
        fprintf(stderr, "\nUsage: bcftools tabix [options] <in.gz> [reg1 [...]]\n\n");
        fprintf(stderr, "Options: -p STR    preset: gff, bed, sam or vcf [gff]\n");
        fprintf(stderr, "         -s INT    column number for sequence names (suppressed by -p) [1]\n");
        fprintf(stderr, "         -b INT    column number for region start [4]\n");
        fprintf(stderr, "         -e INT    column number for region end (if no end, set INT to -b) [5]\n");
        fprintf(stderr, "         -0        specify coordinates are zero-based\n");
        fprintf(stderr, "         -S INT    skip first INT lines [0]\n");
        fprintf(stderr, "         -c CHAR   skip lines starting with CHAR [null]\n");
        fprintf(stderr, "         -a        print all records\n");
        fprintf(stderr, "         -f        force to overwrite existing index\n");
        fprintf(stderr, "         -m INT    set the minimal interval size to 1<<INT; 0 for the old tabix index [0]\n");
        fprintf(stderr, "\n");
        return 1;
    }
    if (is_all) { // read without random access
        kstring_t s;
        BGZF *fp;
        s.l = s.m = 0; s.s = 0;
        fp = bgzf_open(argv[optind], "r");
        while (bgzf_getline(fp, '\n', &s) >= 0) puts(s.s);
        bgzf_close(fp);
        free(s.s);
    } else if (optind + 2 > argc) { // create index
        if ( !conf_ptr )
        {
            // auto-detect file type by file name
            int l = strlen(argv[optind]);
            int strcasecmp(const char *s1, const char *s2);
            if (l>=7 && strcasecmp(argv[optind]+l-7, ".gff.gz") == 0) conf_ptr = &tbx_conf_gff;
            else if (l>=7 && strcasecmp(argv[optind]+l-7, ".bed.gz") == 0) conf_ptr = &tbx_conf_bed;
            else if (l>=7 && strcasecmp(argv[optind]+l-7, ".sam.gz") == 0) conf_ptr = &tbx_conf_sam;
            else if (l>=7 && strcasecmp(argv[optind]+l-7, ".vcf.gz") == 0) conf_ptr = &tbx_conf_vcf;
        }
        if ( conf_ptr ) conf = *conf_ptr;

        if (!is_force) {
            char *fn;
            FILE *fp;
            fn = (char*)alloca(strlen(argv[optind]) + 5);
            strcat(strcpy(fn, argv[optind]), min_shift <= 0? ".tbi" : ".csi");
            if ((fp = fopen(fn, "rb")) != 0) {
                fclose(fp);
                fprintf(stderr, "[E::%s] the index file exists; use option '-f' to overwrite\n", __func__);
                return 1;
            }
        }
        if ( tbx_index_build(argv[optind], min_shift, &conf) )
        {
            fprintf(stderr,"tbx_index_build failed: Is the file bgzip-compressed? Was wrong -p [type] option used?\n");
            return 1;
        }
    } else { // read with random access
        tbx_t *tbx;
        BGZF *fp;
        kstring_t s;
        int i;
        if ((tbx = tbx_index_load(argv[optind])) == 0) return 1;
        if ((fp = bgzf_open(argv[optind], "r")) == 0) return 1;
        s.s = 0; s.l = s.m = 0;
        for (i = optind + 1; i < argc; ++i) {
            hts_itr_t *itr;
            if ((itr = tbx_itr_querys(tbx, argv[i])) == 0) continue;
            while (tbx_bgzf_itr_next(fp, tbx, itr, &s) >= 0) puts(s.s);
            tbx_itr_destroy(itr);
        }
        free(s.s);
        bgzf_close(fp);
        tbx_destroy(tbx);
    }
    return 0;
}
